app/models/batch_connect/session.rb

```diff
@@ -0,0 +1,337 @@
+
+    # Attributes used for serialization
+    # @return [Hash] attributes to be serialized
+    def attributes
+      %w(id cluster_id job_id created_at token title view script_type).map do |attribute|
+        [ attribute, nil ]
+      end.to_h
+    end
+
+    def attributes=(params = {})
+      params.each do |attr, value|
+        self.public_send("#{attr}=", value) if self.respond_to?("#{attr}=")
+      end if params
+    end
+
+    class << self
+      # Root directory for file system database
+      # @return [Pathname] root directory of file system database
+      def db_root
+        OodAppkit.dataroot.join("db").tap { |p| p.mkpath unless p.exist? }
+      end
+
+      # Find all active session jobs
+      # @return [Array<Session>] list of sessions
+      def all
+        db_root.children.select(&:file?).map do |f|
+          new.from_json(f.read)
+        end.map do |s|
+          s.completed? && s.destroy ? nil : s
+        end.compact.sort_by(&:created_at).reverse
+      end
+
+      # Find requested session job
+      # @return [Session] the session
+      def find(id)
+        new.from_json(db_root.join(id).read)
+      end
+    end
+
+    # Path to database file for this object
+    # @return [Pathname, nil] path to db file
+    def db_file
+      self.class.db_root.join(id) if id
+    end
+
+    # Raised when cluster is not found for specific cluster id
+    class ClusterNotFound < StandardError; end
+
+    # Cluster used for controlling job
+    # @raise [ClusterNotFound] if cluster is not found from cluster id
+    # @return [OodCore::Cluster] cluster
+    def cluster
+      OodAppkit.clusters[cluster_id] || raise(ClusterNotFound, "Session specifies nonexistent '#{cluster_id}' cluster id.")
+    end
+
+    # Raised when user doesn't have access to job adapter on cluster
+    class AdapterNotAllowed < StandardError; end
+
+    # Job adapter used for submitting/statusing/deleting jobs
+    # @raise [AdapterNotAllowed] if not allowed to use adapter
+    # @return [OodCore::Job::Adapter, nil] job adapter
+    def adapter
+      cluster.job_allow? ? cluster.job_adapter : raise(AdapterNotAllowed, "Session specifies '#{cluster_id}' cluster id that you do not have access to.")
+    end
+
+    # The batch connect script template used for creating interactive jobs
+    # @param context [#to_h] the context used to render the template
+    # @return [OodCore::BatchConnect::Template] batch connect template
+    def script_template(context = {})
+      cluster.batch_connect_template(template: script_type, **context)
+    end
+
+    # Stage and submit a session from an app and its context
+    # @param app [BatchConnect::App] batch connect app
+    # @param context [BatchConnect::SessionContext] context used for session
+    # @param format [String] format used when rendering template
+    # @return [Boolean] whether saved successfully
+    def save(app:, context:, format: nil)
+      self.id         = SecureRandom.uuid
```

Comment:
We should probably not use SecureRandom.uuid here but do something that ensures there will never be a conflict. Maybe `self.generate_uniq_id`. It could use SecureRandom.uuid but check if there is a conflict.

Conflict is rare but possible because we have different hosts using the same "database". I'll open a separate issue for this though.